File: myersDiff.py
from typing import List, Tuple, Dict

DiffOperation = Tuple[str, int, str]


class DiffNormalizer:
    """Normaliza operações de diff, mesclando operações consecutivas e detectando atualizações."""
    
    @staticmethod
    def normalize(diff_result: List[DiffOperation]) -> List[DiffOperation]:
        """
        Normaliza uma lista de operações de diff.
        
        Args:
            diff_result: Lista de tuplas (operação, posição, caractere)
            
        Returns:
            Lista normalizada de operações de diff
        """
        diff_normalize = list(diff_result)
        
        # Mescla inserções e deleções consecutivas
        DiffNormalizer._merge_consecutive_operations(diff_normalize)
        
        # Detecta atualizações (delete+insert ou insert+delete)
        DiffNormalizer._merge_deletes_and_inserts_as_updates(diff_normalize)
        
        # Mescla atualizações consecutivas
        DiffNormalizer._merge_consecutive_operations(diff_normalize)
        
        return diff_normalize
    
    @staticmethod
    def _merge_consecutive_operations(diff_normalize: List[DiffOperation]) -> None:
        """Mescla operações consecutivas do mesmo tipo e posição adjacente."""
        if len(diff_normalize) <= 1:
            return
        
        index = 0
        while index < len(diff_normalize) - 1:
            op_0, pos_0, char_0 = diff_normalize[index]
            op_1, pos_1, char_1 = diff_normalize[index + 1]
            next_index = pos_0 + len(char_0)
            
            if op_0 == op_1 and next_index == pos_1:
                diff_normalize[index] = (op_0, pos_0, char_0 + char_1)
                del diff_normalize[index + 1]
            else:
                index += 1
    
    @staticmethod
    def _merge_deletes_and_inserts_as_updates(diff_normalize: List[DiffOperation]) -> None:
        """Converte pares delete+insert ou insert+delete em operações de update."""
        if len(diff_normalize) <= 1:
            return
        
        index = 0
        while index < len(diff_normalize) - 1:
            op_0, pos_0, char_0 = diff_normalize[index]
            op_1, pos_1, char_1 = diff_normalize[index + 1]
            next_pos = pos_0 + len(char_0)
            
            # Delete seguido de insert na mesma posição
            if op_0 == 'd' and op_1 == 'i' and next_pos == pos_1:
                diff_normalize[index] = ('u', pos_0, char_1)
                del diff_normalize[index + 1]
            # Insert seguido de delete na mesma posição
            elif op_0 == 'i' and op_1 == 'd' and next_pos == pos_1:
                diff_normalize[index] = ('u', pos_0, char_0)
                del diff_normalize[index + 1]
            else:
                index += 1


class MyersDiff:
    """
    Implementa o algoritmo de Myers para calcular o diff entre duas strings.
    
    O algoritmo de Myers é eficiente para encontrar a menor sequência de edições
    (inserções e deleções) necessárias para transformar uma string em outra.
    """
    
    def __init__(self, original: str, modified: str):
        """
        Inicializa o calculador de diff.
        
        Args:
            original: String original
            modified: String modificada
        """
        self.original = original
        self.modified = modified
        self.n = len(original)
        self.m = len(modified)
        self.max_d = self.n + self.m
        self.v: Dict[int, int] = {1: 0}
        self.trace: List[Dict[int, int]] = []
    
    def compute(self) -> List[DiffOperation]:
        """
        Calcula o diff entre as duas strings.
        
        Returns:
            Lista de tuplas (operação, posição, caractere) onde operação é 'd' (delete), 
            'i' (insert) ou 'u' (update)
        """
        self._find_middle_snake()
        diff_result = self._backtrack()
        diff_result.sort(key=lambda item: item[1])
        return DiffNormalizer.normalize(diff_result)
    
    def _find_middle_snake(self) -> int:
        """
        Encontra a 'snake' do meio usando o algoritmo de Myers.
        
        Returns:
            Número mínimo de edições necessárias
        """
        for d in range(self.max_d + 1):
            self.trace.append(self.v.copy())
            
            for k in range(-d, d + 1, 2):
                # Determina se devemos mover para baixo ou para a direita
                if k == -d or (k != d and self.v.get(k - 1, -1) < self.v.get(k + 1, -1)):
                    x = self.v.get(k + 1, 0)
                else:
                    x = self.v.get(k - 1, 0) + 1
                
                y = x - k
                
                # Segue a diagonal (snake) enquanto os caracteres são iguais
                while x < self.n and y < self.m and self.original[x] == self.modified[y]:
                    x += 1
                    y += 1
                
                self.v[k] = x
                
                # Se chegamos ao fim, encontramos o caminho mais curto
                if x >= self.n and y >= self.m:
                    return d
        
        return self.max_d
    
    def _backtrack(self) -> List[DiffOperation]:
        """
        Reconstrói o caminho de edição a partir do trace.
        
        Returns:
            Lista de operações de diff
        """
        x, y = self.n, self.m
        result = []
        
        for d in range(len(self.trace) - 1, -1, -1):
            v_prev = self.trace[d]
            k = x - y
            
            if k == -d or (k != d and v_prev.get(k - 1, -1) < v_prev.get(k + 1, -1)):
                prev_k = k + 1
            else:
                prev_k = k - 1
            
            prev_x = v_prev.get(prev_k, 0)
            prev_y = prev_x - prev_k
            
            # Snake (diagonal moves - caracteres iguais)
            while x > prev_x and y > prev_y:
                x -= 1
                y -= 1
            
            # Determina se foi uma inserção ou deleção
            if d > 0:
                if x == prev_x:
                    # Movimento vertical = inserção
                    y -= 1
                    result.append(('i', y, self.modified[y]))
                else:
                    # Movimento horizontal = deleção
                    x -= 1
                    result.append(('d', x, self.original[x]))
        
        return list(reversed(result))


class DiffFormatter:
    """Formata resultados de diff em diferentes formatos de saída."""
    
    @staticmethod
    def to_csv(diff_result: List[DiffOperation]) -> str:
        """
        Converte operações de diff para formato CSV.
        
        Args:
            diff_result: Lista de operações de diff
            
        Returns:
            String formatada como CSV
        """
        output = ["op,pos,item"]
        
        for op, pos, char in diff_result:
            if op == 'd':
                output.append(f"-,{pos},{len(char)}")
            elif op == 'i':
                output.append(f"+,{pos},'{char}'")
            else:  # update
                output.append(f"~,{pos},'{char}'")
        
        return '\n'.join(output)


def myers_diff(a: str, b: str) -> List[DiffOperation]:
    """
    Função de conveniência para calcular o diff entre duas strings.
    
    Args:
        a: String original
        b: String modificada
        
    Returns:
        Lista de tuplas (operação, posição, caractere)
    """
    differ = MyersDiff(a, b)
    return differ.compute()

if __name__ == '__main__':
    str_a = "123456"
    str_b = "789878"
    
    result = myers_diff(str_a, str_b)
    print(DiffFormatter.to_csv(result))

File: myersDiff.ts
type DiffOperation = {
  op: 'd' | 'i' | 'u';
  pos: number;
  char: string;
};

/**
 * Normaliza operações de diff, mesclando operações consecutivas e detectando atualizações.
 */
class DiffNormalizer {
  /**
   * Normaliza uma lista de operações de diff.
   */
  static normalize(diffResult: DiffOperation[]): DiffOperation[] {
    const diffNormalize = [...diffResult];
    
    // Mescla inserções e deleções consecutivas
    this.mergeConsecutiveOperations(diffNormalize);
    
    // Detecta atualizações (delete+insert ou insert+delete)
    this.mergeDeletesAndInsertsAsUpdates(diffNormalize);
    
    // Mescla atualizações consecutivas
    this.mergeConsecutiveOperations(diffNormalize);
    
    return diffNormalize;
  }
  
  /**
   * Mescla operações consecutivas do mesmo tipo e posição adjacente.
   */
  private static mergeConsecutiveOperations(diffNormalize: DiffOperation[]): void {
    if (diffNormalize.length <= 1) {
      return;
    }
    
    let index = 0;
    while (index < diffNormalize.length - 1) {
      const { op: op0, pos: pos0, char: char0 } = diffNormalize[index];
      const { op: op1, pos: pos1, char: char1 } = diffNormalize[index + 1];
      const nextIndex = pos0 + char0.length;
      
      if (op0 === op1 && nextIndex === pos1) {
        diffNormalize[index] = { op: op0, pos: pos0, char: char0 + char1 };
        diffNormalize.splice(index + 1, 1);
      } else {
        index++;
      }
    }
  }
  
  /**
   * Converte pares delete+insert ou insert+delete em operações de update.
   */
  private static mergeDeletesAndInsertsAsUpdates(diffNormalize: DiffOperation[]): void {
    if (diffNormalize.length <= 1) {
      return;
    }
    
    let index = 0;
    while (index < diffNormalize.length - 1) {
      const { op: op0, pos: pos0, char: char0 } = diffNormalize[index];
      const { op: op1, pos: pos1, char: char1 } = diffNormalize[index + 1];
      const nextPos = pos0 + char0.length;
      
      // Delete seguido de insert na mesma posição
      if (op0 === 'd' && op1 === 'i' && nextPos === pos1) {
        diffNormalize[index] = { op: 'u', pos: pos0, char: char1 };
        diffNormalize.splice(index + 1, 1);
      }
      // Insert seguido de delete na mesma posição
      else if (op0 === 'i' && op1 === 'd' && nextPos === pos1) {
        diffNormalize[index] = { op: 'u', pos: pos0, char: char0 };
        diffNormalize.splice(index + 1, 1);
      } else {
        index++;
      }
    }
  }
}

/**
 * Implementa o algoritmo de Myers para calcular o diff entre duas strings.
 * 
 * O algoritmo de Myers é eficiente para encontrar a menor sequência de edições
 * (inserções e deleções) necessárias para transformar uma string em outra.
 */
class MyersDiff {
  private original: string;
  private modified: string;
  private n: number;
  private m: number;
  private maxD: number;
  private v: Map<number, number>;
  private trace: Map<number, number>[];
  
  /**
   * Inicializa o calculador de diff.
   */
  constructor(original: string, modified: string) {
    this.original = original;
    this.modified = modified;
    this.n = original.length;
    this.m = modified.length;
    this.maxD = this.n + this.m;
    this.v = new Map([[1, 0]]);
    this.trace = [];
  }
  
  /**
   * Calcula o diff entre as duas strings.
   * 
   * @returns Lista de operações onde operação é 'd' (delete), 'i' (insert) ou 'u' (update)
   */
  compute(): DiffOperation[] {
    this.findMiddleSnake();
    const diffResult = this.backtrack();
    diffResult.sort((a, b) => a.pos - b.pos);
    return DiffNormalizer.normalize(diffResult);
  }
  
  /**
   * Encontra a 'snake' do meio usando o algoritmo de Myers.
   * 
   * @returns Número mínimo de edições necessárias
   */
  private findMiddleSnake(): number {
    for (let d = 0; d <= this.maxD; d++) {
      this.trace.push(new Map(this.v));
      
      for (let k = -d; k <= d; k += 2) {
        // Determina se devemos mover para baixo ou para a direita
        let x: number;
        if (k === -d || (k !== d && (this.v.get(k - 1) ?? -1) < (this.v.get(k + 1) ?? -1))) {
          x = this.v.get(k + 1) ?? 0;
        } else {
          x = (this.v.get(k - 1) ?? 0) + 1;
        }
        
        let y = x - k;
        
        // Segue a diagonal (snake) enquanto os caracteres são iguais
        while (x < this.n && y < this.m && this.original[x] === this.modified[y]) {
          x++;
          y++;
        }
        
        this.v.set(k, x);
        
        // Se chegamos ao fim, encontramos o caminho mais curto
        if (x >= this.n && y >= this.m) {
          return d;
        }
      }
    }
    
    return this.maxD;
  }
  
  /**
   * Reconstrói o caminho de edição a partir do trace.
   * 
   * @returns Lista de operações de diff
   */
  private backtrack(): DiffOperation[] {
    let x = this.n;
    let y = this.m;
    const result: DiffOperation[] = [];
    
    for (let d = this.trace.length - 1; d >= 0; d--) {
      const vPrev = this.trace[d];
      const k = x - y;
      
      let prevK: number;
      if (k === -d || (k !== d && (vPrev.get(k - 1) ?? -1) < (vPrev.get(k + 1) ?? -1))) {
        prevK = k + 1;
      } else {
        prevK = k - 1;
      }
      
      const prevX = vPrev.get(prevK) ?? 0;
      const prevY = prevX - prevK;
      
      // Snake (diagonal moves - caracteres iguais)
      while (x > prevX && y > prevY) {
        x--;
        y--;
      }
      
      // Determina se foi uma inserção ou deleção
      if (d > 0) {
        if (x === prevX) {
          // Movimento vertical = inserção
          y--;
          result.push({ op: 'i', pos: y, char: this.modified[y] });
        } else {
          // Movimento horizontal = deleção
          x--;
          result.push({ op: 'd', pos: x, char: this.original[x] });
        }
      }
    }
    
    return result.reverse();
  }
}

/**
 * Formata resultados de diff em diferentes formatos de saída.
 */
class DiffFormatter {
  /**
   * Converte operações de diff para formato CSV.
   */
  static toCsv(diffResult: DiffOperation[]): string {
    const output: string[] = ['op,pos,item'];
    
    for (const { op, pos, char } of diffResult) {
      if (op === 'd') {
        output.push(`-,${pos},${char.length}`);
      } else if (op === 'i') {
        output.push(`+,${pos},'${char}'`);
      } else {
        // update
        output.push(`~,${pos},'${char}'`);
      }
    }
    
    return output.join('\n');
  }
}

/**
 * Função de conveniência para calcular o diff entre duas strings.
 */
export function myersDiff(a: string, b: string): DiffOperation[] {
  const differ = new MyersDiff(a, b);
  return differ.compute();
}

// Exporta as classes para uso externo
export type { DiffOperation };
export { MyersDiff, DiffNormalizer, DiffFormatter };

// Exemplo de uso (descomente para executar)
// const strA = '123456';
// const strB = '789878';
// const result = myersDiff(strA, strB);
// console.log(DiffFormatter.toCsv(result));
